src/ws/cockpitwsinstancecert.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "cockpitwsinstancecert.h"
+
+#include <assert.h>
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <regex.h>
+#include <stddef.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#define CGROUP_REGEX         "^(0:|1:name=systemd):/system.slice/system-cockpithttps.slice/" \
+                             "cockpit-wsinstance-https@([0-9a-f]{64}).service$"
+#define CGROUP_REGEX_FLAGS   (REG_EXTENDED | REG_NEWLINE)
+#define CGROUP_REGEX_GROUPS  3   /* number of groups, including the complete match */
+#define CGROUP_REGEX_MATCH   2   /* the group which contains the instance */
+
+/* get our cgroup, map it to a systemd unit instance name
+ * looks like 0::/system.slice/system-cockpit\x2dwsinstance\x2dhttps.slice/cockpit-wsinstance-https@123abc.service
+ * returns "123abc" instance name (static string)
+ */
+static const char *
+get_ws_https_instance (void)
```

Comment:
Using the cgroup for this information is counter to the typical flow for Cockpit. The typical flow is:

 1. cockpit-ws: relay credentials from untrusted source
 2. cockpit-session: Receive credentials from untrusted source
 3. cockpit-session, PAM: Use credentials to authenticate
 4. cockpit-session: Use authentication to start session

To follow along this would be:

 1. cockpit-ws: Relay certificate information received in "authorize" tls-cert message
 2. cockpit-session: Parse out that tls-cert and pass to PAM module as a auth token
 3. pam_cockpit_cert.so: Check that auth token is actually the certificate used with the connection, check it matches the user.

Reply:
Talked about this on IRC. Making this conform to the typical flow is just extra needless work.

Reply:
We discussed this at length on IRC, but summary: this "typical" information flow doesn't work for this setup. The whole point of splitting out TLS termination and isolation ws instances for different certificates is that we assume ws can be hacked, and thus can't give us trustworthy information. Passing something like a sealed memfd would be okayish (although ws can try to pass it on to a parallel ws instance, at which point it can impersonate another user session again). Thus c-session/the PAM module have to check the cgroup (which really means "which certificate does this ws instance handle?") anyway, and at that point there is no further information that ws could give to PAM that it doesn't already possess.

Lis had an interesting idea with tls creating nonces for each connection, sending these to ws as the first thing after `accept()`, and then passing on these nonces through the "tls-cert" auth  protocol. That would again be more in line with the "typical flow".